kuma/wiki/feeds.py

```diff
@@ -52,7 +53,33 @@ def get_object(self, request, format):
             self.feed_type = Atom1Feed
 
     def item_pubdate(self, document):
-        return document.current_revision.created
+        """
+        Force pubdate to be timezone-aware.
+
+        When switching to standard time in the second half of the year, an hour
+        occurs twice. In the United States, 2:00 am becomes 1:00 am again.
+        The naive datetimes stored in our database can't distinguish beteween
+        the first and second time the hour occurs. An error is raised, and no
+        usable feed returned.
+
+        This handler assumes that it is the first hour, so it will be wrong 50%
+        of the time, but this is better than an errored feed. We can do better
+        when we convert to storing timezone-aware datetimes, by assuming
+        revision IDs increase and finding the transition point.
+
+        https://stackoverflow.com/questions/21465528/resolving-ambiguoustimeerror-from-djangos-make-aware
+        TODO: In Django 1.9, can pass is_dst=False to make_aware
+        TODO: Enable timezone-aware dates in database, remove check
+        """
+        raw_pubdate = document.current_revision.created
+        if is_naive(raw_pubdate):
+            # USE_TZ is False, database has naive timestamps
+            timezone = get_current_timezone()
+            pubdate = timezone.localize(raw_pubdate, is_dst=True)
```

Comment:
I could not find any documentation of this `localize` in documentation.
by supplying `is_dst=True`, are we assuming that its always Daylight Saving time?

Reply:
@safwanrahman Yes, we're just going to assume it's always DST until we start storing timezone-aware datetimes.

Reply:
@safwanrahman I had similar questions that were not answered by the docs I wrote. I assumed it was OK because other tests outside the magic hour passed. I'm investigating a little more now that you've asked it. Results to come...

Reply:
``localize`` is provided by pytz, which has [pretty good narrative documentation](http://pytz.sourceforge.net), but doesn't go into details about the functions.

The details are in the source, hosted on launchpad:

https://git.launchpad.net/pytz/tree/src/pytz/tzinfo.py#n244

The code looks forward a day and back a day, to see if the time is around a transition. This can result in 0, 1, or 2 possible timezones:

- 0 cases: The time is an "impossible time", during the lost hour during the transition into daylight savings time. ``is_dst=None`` raises an error, ``is_dst=True`` uses the timezone 6 hours later, and ``is_dst=False`` uses the timezone 6 hours earlier.
- 1 case: There is no ambiguity, the one timezone is used
- 2 cases: The time is from the transition out of daylight savings time, when an hour is repeated. ``is_dst=None`` raises an error, ``is_dst=True`` uses the DST timezone, ``is_dst=False`` uses the standard timezone.

I'm not Python, so I could have read the code wrong. Here's some test code in the Django shell:

```
>>> from datetime import datetime as dt
>>> from django.utils.timezone import get_current_timezone
>>> for month in range(1,13):
...   test = dt(2017, month, 3, 1, 23)
...   print(get_current_timezone().localize(test))
...
2017-01-03 01:23:00-08:00
2017-02-03 01:23:00-08:00
2017-03-03 01:23:00-08:00
2017-04-03 01:23:00-07:00
2017-05-03 01:23:00-07:00
2017-06-03 01:23:00-07:00
2017-07-03 01:23:00-07:00
2017-08-03 01:23:00-07:00
2017-09-03 01:23:00-07:00
2017-10-03 01:23:00-07:00
2017-11-03 01:23:00-07:00
2017-12-03 01:23:00-08:00
```
Here in the Central time zone, the dates from April through November are Central Daylight Time (-07:00), and the other months are Central Standard Time (-8:00).

That was a fun question, thanks!